summarize/src/signed_duration.rs

```diff
@@ -0,0 +1,118 @@
+use std::cmp::{Ord, Ordering, PartialOrd};
+use std::fmt;
+use std::ops::Sub;
+use std::time::Duration;
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Clone, Copy, Eq, PartialEq)]
+pub struct SignedDuration {
+    pub duration: Duration,
+    pub is_positive: bool,
+}
+
+impl From<Duration> for SignedDuration {
+    fn from(d: Duration) -> SignedDuration {
+        SignedDuration {
+            duration: d,
+            is_positive: true
+        }
+    }
+}
+
+impl Ord for SignedDuration {
+    fn cmp(&self, other: &SignedDuration) -> Ordering {
+        match (self.is_positive, other.is_positive) {
+            (true, true) => self.duration.cmp(&other.duration),
+            (false, false) => other.duration.cmp(&self.duration),
+            (true, false) => Ordering::Greater,
+            (false, true) => Ordering::Less,
+        }
+    }
+}
+
+impl PartialOrd for SignedDuration {
+    fn partial_cmp(&self, other: &SignedDuration) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Sub for SignedDuration {
+    type Output = SignedDuration;
+
+    fn sub(self, rhs: SignedDuration) -> SignedDuration {
+        let (result, result_is_positive) =
+            match (self.is_positive, rhs.is_positive) {
+                (true, true) if self.duration >= rhs.duration =>
+                    (self.duration - rhs.duration, true),
+                (true, true) => (rhs.duration - self.duration, false),
+
+                (false, true) => (self.duration + rhs.duration, false),
+                (true, false) => (self.duration + rhs.duration, true),
+
+                (false, false) if self.duration > rhs.duration =>
+                    (self.duration - rhs.duration, false),
+                (false, false) => (rhs.duration - self.duration, true),
+            };
```

Comment:
I think this would be a lot easier to interpret if we convert to nanoseconds and then back into a Duration.